File: kelly/linked-list.ts
class Matrix<T> {
  private matrix: T[][];

  constructor(rows: number, cols: number) {
    this.matrix = [];
    for (let i = 0; i < rows; i++) {
      this.matrix[i] = new Array(cols);
    }
  }

  insert(item: T, i: number, j: number): void {
    if (i >= 0 && i < this.matrix.length && j >= 0 && j < this.matrix[0].length) {
      this.matrix[i][j] = item;
    }
  }

  traverse(): void {
    for (let i = 0; i < this.matrix.length; i++) {
      for (let j = 0; j < this.matrix[0].length; j++) {
        console.log(this.matrix[i][j]);
      }
    }
  }

  search(item: T): boolean {
    for (let i = 0; i < this.matrix.length; i++) {
      for (let j = 0; j < this.matrix[0].length; j++) {
        if (this.matrix[i][j] === item) {
          return true;
        }
      }
    }
    return false;
  }

  getIndex(item: T): { i: number, j: number } | null {
    for (let i = 0; i < this.matrix.length; i++) {
      for (let j = 0; j < this.matrix[0].length; j++) {
        if (this.matrix[i][j] === item) {
          return { i, j };
        }
      }
    }
    return null;
  }

  reverseRow(rowIndex: number): void {
    if (rowIndex >= 0 && rowIndex < this.matrix.length) {
      this.matrix[rowIndex].reverse();
    }
  }

  reverseColumn(columnIndex: number): void {
    if (columnIndex >= 0 && columnIndex < this.matrix[0].length) {
      const column = [];
      for (let i = 0; i < this.matrix.length; i++) {
        column.push(this.matrix[i][columnIndex]);
      }
      column.reverse();
      for (let i = 0; i < this.matrix.length; i++) {
        this.matrix[i][columnIndex] = column[i];
      }
    }
  }

  sort(item: T): void {
    for (let i = 0; i < this.matrix.length; i++) {
      for (let j = 0; j < this.matrix[0].length; j++) {
        if (this.matrix[i][j] === item) {
          this.matrix[i].splice(j, 1);
          break;
        }
      }
    }
  }

  diagonalSum(): number {
    let sum = 0;
    const minSize = Math.min(this.matrix.length, this.matrix[0].length);
    for (let i = 0; i < minSize; i++) {
      sum += this.matrix[i][i];
    }
    return sum;
  }

  display(): void {
    for (let i = 0; i < this.matrix.length; i++) {
      console.log(this.matrix[i].join('\t'));
    }
  }
}


const matrix = new Matrix<number>(3, 3);

matrix.insert(1, 0, 0);
matrix.insert(2, 0, 1);
matrix.insert(3, 0, 2);
matrix.insert(4, 1, 0);
matrix.insert(5, 1, 1);
matrix.insert(6, 1, 2);
matrix.insert(7, 2, 0);
matrix.insert(8, 2, 1);
matrix.insert(9, 2, 2);

matrix.display(); // Display the matrix
// Output:
// 1    2    3
// 4    5    6
// 7    8    9

matrix.traverse(); // Traverse and print every cell value

console.log(matrix.search(5)); // true

console.log(matrix.getIndex(6)); // { i: 1, j: 2 }

matrix.reverseRow(1);
matrix.display();
// Output:
// 1    2    3
// 6    5    4
// 7    8    9

matrix.reverseColumn(0);
matrix.display();
// Output:
// 3    2    1
// 6    5    4
// 9    8    7

matrix.sort(5);
matrix.display();
// Output:
// 3    2    1
// 6    4
// 9    8    7

console.log(matrix.diagonalSum()); // 14

File: kelly/linked-list.cs
using System;

class Matrix
{
    private int[,] data;

    public Matrix(int rows, int columns)
    {
        data = new int[rows, columns];
    }

    public void Insert(int item, int i, int j)
    {
        data[i, j] = item;
    }

    public void Traverse()
    {
        int rows = data.GetLength(0);
        int columns = data.GetLength(1);

        for (int i = 0; i < rows; i++)
        {
            for (int j = 0; j < columns; j++)
            {
                Console.Write(data[i, j] + " ");
            }
            Console.WriteLine();
        }
    }

    public bool Search(int item)
    {
        int rows = data.GetLength(0);
        int columns = data.GetLength(1);

        for (int i = 0; i < rows; i++)
        {
            for (int j = 0; j < columns; j++)
            {
                if (data[i, j] == item)
                {
                    return true;
                }
            }
        }

        return false;
    }

    public Tuple<int, int> GetIndex(int item)
    {
        int rows = data.GetLength(0);
        int columns = data.GetLength(1);

        for (int i = 0; i < rows; i++)
        {
            for (int j = 0; j < columns; j++)
            {
                if (data[i, j] == item)
                {
                    return Tuple.Create(i, j);
                }
            }
        }

        return null;
    }

    public void ReverseRow(int rowIndex)
    {
        int columns = data.GetLength(1);
        int start = 0;
        int end = columns - 1;

        while (start < end)
        {
            int temp = data[rowIndex, start];
            data[rowIndex, start] = data[rowIndex, end];
            data[rowIndex, end] = temp;

            start++;
            end--;
        }
    }

    public void ReverseColumn(int columnIndex)
    {
        int rows = data.GetLength(0);
        int start = 0;
        int end = rows - 1;

        while (start < end)
        {
            int temp = data[start, columnIndex];
            data[start, columnIndex] = data[end, columnIndex];
            data[end, columnIndex] = temp;

            start++;
            end--;
        }
    }

    public void Sort(int item)
    {
        int rows = data.GetLength(0);
        int columns = data.GetLength(1);

        for (int i = 0; i < rows; i++)
        {
            for (int j = 0; j < columns; j++)
            {
                if (data[i, j] == item)
                {
                    data[i, j] = 0; // Assuming "removing" by setting to 0
                }
            }
        }
    }

    public int DiagonalSum()
    {
        int rows = data.GetLength(0);
        int columns = data.GetLength(1);
        int sum = 0;

        if (rows == columns)
        {
            for (int i = 0; i < rows; i++)
            {
                sum += data[i, i];
            }
        }

        return sum;
    }

    public void Display()
    {
        Traverse(); // Simply printing the matrix using Traverse method
    }
}

class Program
{
    static void Main(string[] args)
    {
        Matrix matrix = new Matrix(3, 4);

        matrix.Insert(1, 0, 0);
        matrix.Insert(2, 0, 1);
        matrix.Insert(3, 0, 2);
        matrix.Insert(4, 0, 3);
        matrix.Insert(5, 1, 0);
        matrix.Insert(6, 1, 1);
        matrix.Insert(7, 1, 2);
        matrix.Insert(8, 1, 3);
        matrix.Insert(9, 2, 0);
        matrix.Insert(10, 2, 1);
        matrix.Insert(11, 2, 2);
        matrix.Insert(12, 2, 3);

        Console.WriteLine("Matrix:");
        matrix.Traverse();

        Console.WriteLine("Search for item 6: " + matrix.Search(6));
        Console.WriteLine("Search for item 13: " + matrix.Search(13));

        Tuple<int, int> index = matrix.GetIndex(7);
        Console.WriteLine("Index of item 7: (" + index.Item1 + ", " + index.Item2 + ")");

        matrix.ReverseRow(1);
        Console.WriteLine("Matrix after reversing row 1:");
        matrix.Traverse();

        matrix.ReverseColumn(2);
        Console.WriteLine("Matrix after reversing column 2:");
        matrix.Traverse();

        matrix.Sort(8);
        Console.WriteLine("Matrix after sorting (removing) item 8:");
        matrix.Traverse();

        Console.WriteLine("Sum of diagonal elements: " + matrix.DiagonalSum());

        Console.WriteLine("Matrix displayed using Display method:");
        matrix.Display();

        Console.ReadLine();
    }
}
